### PR TITLE
Strongly connected components

### PR DESCRIPTION
**Pull request by nikalosa**
### **Describe your change:**
Algorithm for finding out strongly connected components in a graph

* Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* This pull request is all my own work -- I have not plagiarized.
* I know that pull requests will not be merged if they fail the automated tests.
* This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* All new Python files are placed inside an existing directory.
* All filenames are in all lowercase characters with no spaces or dashes.
* All functions and variable names follow Python naming conventions.
* All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: graphs/strongly_connected_components.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    0: [1, 2, 3],
+    1: [2],
+    2: [0],
+    3: [4],
+    4: [5],
+    5: [3],
+}
+
+
+def topology_sort(graph, vert, visited, order):
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:26]: Please add typehints and doctests

File: graphs/strongly_connected_components.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    visited[vert] = True
+
+    for neighbour in graph[vert]:
+        if not visited[neighbour]:
+            topology_sort(graph, neighbour, visited, order)
+
+    order.append(vert)
+
+
+def find_components(reverse_graph, vert, visited, component):
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:41]: Please add typehints and doctests

**Comment by mateuszz0000** [comment]
Please fix also flake8 errors: https://travis-ci.com/github/TheAlgorithms/Python/builds/171253892

File: graphs/strongly_connected_components.py (hunk 1 above)
Comment (nikalosa) [reply graphs/strongly_connected_components.py]: Done!

File: graphs/strongly_connected_components.py (hunk 2 above)
Comment (nikalosa) [reply graphs/strongly_connected_components.py]: Done!

**Comment by nikalosa** [comment]
> Please fix also flake8 errors: https://travis-ci.com/github/TheAlgorithms/Python/builds/171253892

That error already has fixed after the first commit.

File: graphs/strongly_connected_components.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    At this time graph is the same as input
+    >>> topology_sort(test_graph_1, 0, 5 * [False])
+    [1, 2, 4, 3, 0]
+    >>> topology_sort(test_graph_2, 0, 6 * [False])
+    [2, 1, 5, 4, 3, 0]
+    """
+
+    visited[vert] = True
+    order = []
+
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:38]: `visited[vert], order = True, [] `

File: graphs/strongly_connected_components.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    >>> find_components({0: [1], 1: [2], 2: [0]}, 0, 5 * [False])
+    [0, 1, 2]
+    >>> find_components({0: [2], 1: [0], 2: [0, 1]}, 0, 6 * [False])
+    [0, 2, 1]
+    """
+
+    visited[vert] = True
+
+    component = [vert]
+
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:61]: `visited[vert], component = True, [vert]`

File: graphs/strongly_connected_components.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    This function takes graph as a parameter
+    and then returns the list of strongly connected components
+    >>> strongly_connected_components(test_graph_1)
+    [[0, 1, 2], [3], [4]]
+    >>> strongly_connected_components(test_graph_2)
+    [[0, 2, 1], [3, 5, 4]]
+    """
+
+    n = len(graph)
+
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:80]: `graph_size` could be more readable instead `n`

File: graphs/strongly_connected_components.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    visited = n * [False]
+
+    # describe reversed graph (All edges reversed)
+    reverse_graph = {vert: [] for vert in range(n)}
+
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:85]: `reverse_graph` --> `reversed_graph` . I think comment above is unnecessary. Variable name defines exatly what is it.

File: graphs/strongly_connected_components.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    for i in range(n):
+        if not visited[i]:
+            order += topology_sort(graph, i, visited)
+            # print(order, 'aeee\n')
+
+    # answer list
+    components_list = []
+    # reinitialize visited list
+    visited = n * [False]
+
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:102]: Remove comments, rename variables better.

File: graphs/strongly_connected_components.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            components_list.append(component)
+
+    return components_list
+
+
+if __name__ == "__main__":
+
+    import doctest
+
+    doctest.testmod()
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:117]: I think here should be simple example to run. Doctests are running via pytest.

File: graphs/strongly_connected_components.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
+"""
+https://en.wikipedia.org/wiki/Strongly_connected_component
+
+Finding strongly connected components in directed graph
+
+"""
+
+test_graph_1 = {
```
Comment (mateuszz0000) [inline graphs/strongly_connected_components.py:8]: @cclauss Do you have any idea how we can avoid declaring these variables here but keep current doctests?

File: graphs/strongly_connected_components.py (hunk 3 above)
Comment (cclauss) [reply graphs/strongly_connected_components.py:38]: I find the current code more readable than the proposed code.

File: graphs/strongly_connected_components.py (hunk 4 above)
Comment (cclauss) [reply graphs/strongly_connected_components.py:61]: I find the current code more readable than the proposed code but we could remove the blank line between the two statements.

File: graphs/strongly_connected_components.py (hunk 8 above)
Comment (cclauss) [reply graphs/strongly_connected_components.py:117]: This is an interesting point.  I could go either way but I kinda like that the program autoruns the test suite _before_ it accepts the user's workload.  Kinda _better safe than sorry_ despite being a bit more verbose and unoptimized.  Optimized algorithms is a useful goal but optimized `__main__` is far less important.

File: graphs/strongly_connected_components.py (hunk 9 above)
Comment (cclauss) [reply graphs/strongly_connected_components.py:8]: I think we are OK here.  Naming helps a lot so when I see `test_graph_1`, I immediately know that this data is for testing purposes.  If I was going to move this algorithm to some larger program, I will need to do some level of refactoring.  I might move all the doctests and the test data to a separate test_xyz.py file or I might remove them altogether.  For our purposes in this repo, a single file that contains the algorithm, tests, and test data is good for learning purposes as long as we are diligent with our naming.

File: graphs/strongly_connected_components.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    Use depth first search to find strongliy connected
+    vertices. Now graph is reversed
+    >>> find_components({0: [1], 1: [2], 2: [0]}, 0, 5 * [False])
+    [0, 1, 2]
+    >>> find_components({0: [2], 1: [0], 2: [0, 1]}, 0, 6 * [False])
+    [0, 2, 1]
+    """
+
+    visited[vert] = True
+
```
Comment (cclauss) [inline graphs/strongly_connected_components.py:59]:
```suggestion
```

File: graphs/strongly_connected_components.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    and then returns the list of strongly connected components
+    >>> strongly_connected_components(test_graph_1)
+    [[0, 1, 2], [3], [4]]
+    >>> strongly_connected_components(test_graph_2)
+    [[0, 2, 1], [3, 5, 4]]
+    """
+
+    n = len(graph)
+
+    visited = n * [False]
```
Comment (cclauss) [inline graphs/strongly_connected_components.py:81]:
```suggestion
    visited = len(graph) * [False]
```

File: graphs/strongly_connected_components.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    reversed_graph = {vert: [] for vert in range(len(graph))}
+
+    for vert, neighbours in graph.items():
+        for neighbour in neighbours:
+            reversed_graph[neighbour].append(vert)
+
+    order = []
+    for i in range(len(graph)):
+        if not visited[i]:
+            order += topology_sort(graph, i, visited)
```
Comment (cclauss) [inline graphs/strongly_connected_components.py:88]: ___OPTIONAL:___  For me `range(len(graph))` is always a sign that perhaps I should consider [`enumerate()`](https://docs.python.org/3/library/functions.html#enumerate) instead.  Sometimes this helps understand-ability.
```suggestion
    for i, was_visited in enumerate(graph):
        if not was_visited:
            order += topology_sort(graph, i, visited)
```

Comment (nikalosa) [reply graphs/strongly_connected_components.py:88]: Great point, I will change that right away, thanks

**Pull request merged**